### PR TITLE
transform: add `ColumnNames` attribute

### PR DESCRIPTION
**Pull request by aalexandrov**
Add a `ColumnNames` attribute and expose it in `EXPLAIN` outputs through the `column_names` option.

### Motivation

  * This PR adds a feature that has not yet been specified.

We don't have an issue for that, but multiple people have expressed interest in having an `EXPLAIN` plan that prints humanized column names. This PR is the first step towards that.

### Tips for reviewer

See commit messages for details (specifically the message of the third commit for motivation and background of the refactoring).

The first four commits do some refactoring to the testing and attribute derivation frameworks required that is needed in order to implement the `ColumnNames` inference logic. More specifically, we need a way to create `Attribute` implementations with parameters (until now the structs implementing `Attribute` were all nullary). This is achieved by the third and fourth commits which:

- Introduce a struct-based `DerivedAttributes` container.
- Pass context to the attribute constructors.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/transform/src/attribute/column_names.rs [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+
+                for (i, base_column_name) in base_results.iter().enumerate() {
+                    let mut column_name = base_column_name.clone();
+                    for input_results in inputs_results.iter() {
+                        if !input_results[i].is_empty() {
+                            column_name = input_results[i].clone();
+                            break;
+                        }
+                    }
+                    column_names.push(column_name);
```
Comment (vmarcos) [inline src/transform/src/attribute/column_names.rs:196]: Not sure if I understand the MIR `Union` semantics 100% vs. the logic here: If we have a `base`, then wouldn't the column names there always be ignored if there are other inputs with column names present? Is that intended?

For example, consider a test case in which a `Union` would have a `base` and two `inputs` s.t. the `base` and the last of `inputs` have all column names given. Wouldn't the fragment above take the column names from the last of `inputs` and not from `base`?

Comment (aalexandrov) [reply src/transform/src/attribute/column_names.rs:196]: Yes, currently the last input wins. I can change this to "first input wins" semantics, but most of the time I am expecting identical input positions to also resolve to identical column names (at least in SQL they are required to align). 

I don't think there is a silver bullet solution here. Consider the following queries:

```sql
CREATE TABLE R(x int, y int);
CREATE TABLE S(y int, x int);

-- Q1
EXPLAIN WITH(column_names)
SELECT x as y, y as x FROM R 
UNION ALL
SELECT y, x FROM S;

-- Q2
EXPLAIN WITH(column_names)
SELECT x, y FROM R 
UNION ALL
SELECT y as x, x as y FROM S;
```

Currently, in both cases we will infer `(y, x)`, which is bad for Q2 where the schema is actually `(x, y)`. If I change the behavior it in this degenerate case things will get bad for Q1, but I think it's probably the more intuitive solution.

File: src/transform/src/attribute/mod.rs [hunk 2]
```diff
@@ -256,10 +250,12 @@ impl RequiredAttributes {
...
 /// from an [ExplainConfig].
 pub struct DerivedAttributes {
-    attributes: Box<TypeMap>,
-    /// Holds attributes that have not yet been derived for a given
-    /// subexpression.
-    to_be_derived: Box<TypeMap>,
+    /// Holds the attributes derived for all subexpressions so far.
+    store: Box<AttributeStore>,
+    /// A buffer to be used for the next subexpresion. For every subexpressions
+    /// we swap the `store` and `buffer` state and then process iterms from
```
[the remark this answers is not in the document]
Comment (aalexandrov) [reply src/transform/src/attribute/mod.rs:266]: Freudian slip, I guess I'm missing the iTerm2 experiene.

File: src/transform/src/attribute/column_names.rs [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+            Get {
+                id: Id::Local(id),
+                typ,
+                access_strategy: _,
+            } => {
+                if let Some(column_names) = self.env.get(id) {
+                    column_names.clone()
+                } else {
+                    // TODO: this really shouldn't happen
+                    ColumnNames::anonymous(0..typ.arity()).collect()
```
[the remark this answers is not in the document]
Comment (aalexandrov) [reply src/transform/src/attribute/column_names.rs:73]: Done.

File: src/transform/src/attribute/column_names.rs (hunk 1 above)
Comment (vmarcos) [reply src/transform/src/attribute/column_names.rs:196]: What I'm finding confusing is that actually it is not that the last input wins, but the first with names that is different than `base`. Let me augment your example a bit:
```sql
CREATE TABLE R(x int, y int);
CREATE TABLE S(y int, x int);
CREATE TABLE T(a int, b int);
CREATE TABLE U(c int, d int);

-- Q1
EXPLAIN WITH(column_names)
SELECT x as y, y as x FROM R 
UNION ALL
SELECT y, x FROM S;

-- Q2
EXPLAIN WITH(column_names)
SELECT x, y FROM R 
UNION ALL
SELECT y as x, x as y FROM S;

-- Q3
EXPLAIN WITH(column_names)
SELECT x as y, y as x FROM R 
UNION ALL
SELECT y, x FROM S
UNION ALL
SELECT a, b FROM T;

-- Q4
EXPLAIN WITH(column_names)
SELECT x, y FROM R 
UNION ALL
SELECT a, b FROM T
UNION ALL
SELECT y as x, x as y FROM S;

-- Q5
EXPLAIN WITH(column_names)
SELECT x, y FROM R 
UNION ALL
SELECT a + 1, b + 1 FROM T
UNION ALL
SELECT c, d FROM U
UNION ALL
SELECT y as x, x as y FROM S;
```
What we get as a result for the `EXPLAIN`s is:
```
                           Optimized Plan                           
--------------------------------------------------------------------
 Explained Query:                                                  +
   Union // { column_names: "(y, x)" }                             +
     ReadStorage materialize.public.r // { column_names: "(x, y)" }+
     ReadStorage materialize.public.s // { column_names: "(y, x)" }+
 
(1 row)

                           Optimized Plan                           
--------------------------------------------------------------------
 Explained Query:                                                  +
   Union // { column_names: "(y, x)" }                             +
     ReadStorage materialize.public.r // { column_names: "(x, y)" }+
     ReadStorage materialize.public.s // { column_names: "(y, x)" }+
 
(1 row)

                           Optimized Plan                           
--------------------------------------------------------------------
 Explained Query:                                                  +
   Union // { column_names: "(y, x)" }                             +
     ReadStorage materialize.public.r // { column_names: "(x, y)" }+
     ReadStorage materialize.public.s // { column_names: "(y, x)" }+
     ReadStorage materialize.public.t // { column_names: "(a, b)" }+
 
(1 row)

                           Optimized Plan                           
--------------------------------------------------------------------
 Explained Query:                                                  +
   Union // { column_names: "(a, b)" }                             +
     ReadStorage materialize.public.r // { column_names: "(x, y)" }+
     ReadStorage materialize.public.t // { column_names: "(a, b)" }+
     ReadStorage materialize.public.s // { column_names: "(y, x)" }+
 
(1 row)

                             Optimized Plan                             
------------------------------------------------------------------------
 Explained Query:                                                      +
   Union // { column_names: "(c, d)" }                                 +
     ReadStorage materialize.public.r // { column_names: "(x, y)" }    +
     Project (#2, #3) // { column_names: "(#0, #1)" }                  +
       Map ((#0 + 1), (#1 + 1)) // { column_names: "(a, b, #2, #3)" }  +
         ReadStorage materialize.public.t // { column_names: "(a, b)" }+
     ReadStorage materialize.public.u // { column_names: "(c, d)" }    +
     ReadStorage materialize.public.s // { column_names: "(y, x)" }    +
 
(1 row)
```

File: src/transform/src/attribute/column_names.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (aalexandrov) [reply src/transform/src/attribute/column_names.rs]: I tried adding the panic and unfortunately discovered situations in `LetRec` blocks where unfortunately this _could_ happen. I changed the comment to reflect that.

File: src/transform/src/attribute/column_names.rs (hunk 1 above)
Comment (aalexandrov) [reply src/transform/src/attribute/column_names.rs]: I change that code to reflect what the comment says:

```rust
// Use the first non-empty column across all inputs.
let mut column_names = vec![];

let base_results = input_results.next().unwrap();
let inputs_results = input_results.collect::<Vec<_>>();

for (i, mut column_name) in base_results.iter().cloned().enumerate() {
    for input_results in inputs_results.iter() {
        if column_name.is_empty() && !input_results[i].is_empty() {
            column_name = input_results[i].clone();
            break;
        }
    }
    column_names.push(column_name);
}
```

Comment (vmarcos) [reply src/transform/src/attribute/column_names.rs]: Perfect, thanks!

**Review by vmarcos — APPROVED** [review]
LGTM!

**Pull request merged**